From 06fcb28c303ee4d2dddb0f4ea39037b163f9e21f Mon Sep 17 00:00:00 2001
From: Alexey Turitsyn <alexey.turitsyn@lge.com>
Date: Wed, 20 May 2015 14:05:45 +0300
Subject: [PATCH] Fix-HarfBuzz-NG-regression

:Release Notes:
Patch defines list of explicit Unicode scripts and assign
"Common" script property to other values that are not in that list

:Detailed Notes:
Starting from Qt5.4, the HarfBuzz-NG library is used by default and it gives regression (some chars do not dispalyed)
Regression happens because of incorrect Unicode script property value (http://www.unicode.org/reports/tr24/)
This patch defines list of explicit Unicode scripts (being copied from old Harfbuzz library) and assign
"Common" script property to other values that are not in that list

:Testing Performed:
webos-pro/master/816

:QA Notes:

:Issues Addressed:
[DRD-7977] Re-enable use of harfbuzz-ng

Upstream-Status: Inappropriate[bugfix for DRD-6125]
---
 src/gui/text/qtextengine.cpp | 39 +++++++++++++++++++++++++++++++--------
 1 file changed, 31 insertions(+), 8 deletions(-)

diff --git a/src/gui/text/qtextengine.cpp b/src/gui/text/qtextengine.cpp
index f569c6e..85fd2d2 100644
--- a/src/gui/text/qtextengine.cpp
+++ b/src/gui/text/qtextengine.cpp
@@ -1675,16 +1675,39 @@ void QTextEngine::itemize() const
         // ### pretend HB-old behavior for now
         for (int i = 0; i < length; ++i) {
             switch (analysis[i].script) {
-            case QChar::Script_Latin:
-            case QChar::Script_Hiragana:
-            case QChar::Script_Katakana:
-            case QChar::Script_Bopomofo:
-            case QChar::Script_Han:
-                analysis[i].script = QChar::Script_Common;
+            case QChar::Script_Inherited:
+            case QChar::Script_Common:
+            case QChar::Script_Arabic:
+            case QChar::Script_Armenian:
+            case QChar::Script_Bengali:
+            case QChar::Script_Cyrillic:
+            case QChar::Script_Devanagari:
+            case QChar::Script_Georgian:
+            case QChar::Script_Greek:
+            case QChar::Script_Gujarati:
+            case QChar::Script_Gurmukhi:
+            case QChar::Script_Hangul:
+            case QChar::Script_Hebrew:
+            case QChar::Script_Kannada:
+            case QChar::Script_Khmer:
+            case QChar::Script_Lao:
+            case QChar::Script_Malayalam:
+            case QChar::Script_Myanmar:
+            case QChar::Script_Ogham:
+            case QChar::Script_Oriya:
+            case QChar::Script_Runic:
+            case QChar::Script_Sinhala:
+            case QChar::Script_Syriac:
+            case QChar::Script_Tamil:
+            case QChar::Script_Telugu:
+            case QChar::Script_Thaana:
+            case QChar::Script_Thai:
+            case QChar::Script_Tibetan:
+            case QChar::Script_Nko:
                 break;
             default:
-                break;
-            }
+                analysis[i].script = QChar::Script_Common;
+            };
         }
     } else {
         for (int i = 0; i < length; ++i)
